### ISSUE TITLE
Get statistics about timing sections across different processors

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#1139]
Scalability to large numbers of processors is disproportionately affected by processors that have _more_ work than the average, as they keep on hold a potentially very large number of other processors while they finish up their work.
#1131 now provides a facility where we can rebalance a mesh if it turns out that some processors repeatedly have more work to do than others. We could use this to make smarter decisions regarding balancing. However, we lack any way to test which processors take long and which don't.

What we would need is a way to get statistics across processors how long operations took. One way I see would be to extend the `TimerOutput` class to not only get overall wallclock time for each section, but to also provide per-processor wallclock time. To make this accurate, we would probably have to run a barrier also upon entering a section, in addition to exiting a section. I have no idea what we would find if we plotted the times each processor takes for each section of, say, one timestep of ASPECT, but I imagine we would learn something from it.

**Comment by kronbichler** [comment]
I like the idea. However, I would not expect too much in terms of revealing all imbalance bottlenecks (FWIR about the timer structure in ASPECT). The reason is communication that is done in different phases.

I did a detailed check for the analysis of one of my programs with a solver similar to the Stokes solver in ASPECT/step-32. The solver uses the following blocks:
- Matrix-vector product
- Application of ILU on velocity variables
- Multiplication by divergence matrix
- Application of ILU on pressure variable
- Application of AMG on pressure variable
- Orthogonalization work in GMRES

I introduced timers both around the whole solver as well as inside the matrix classes and the block diagonal preconditioner. Here are some timings on 16,384 cores (min and max over all processors):
```
- Total solver time: 11.47s ... 11.52s
- Matrix-vector product: 1.29s ... 1.97s
- Velocity ILU: 0.919s ... 1.44s
- Divergence matrix: 0.378s ... 0.861s
- Pressure ILU: 0.0351s ... 0.0568s
- Pressure AMG: 7.37s ... 7.97s
- Total preconditioner time: 9.46s ... 9.62s
```
The interesting part are the operations without communication (ILU without overlap) which vary quite a lot on the given example because some processors have a lot of boundary dofs where there are fewer matrix entries. The processor reporting the minimum run time for the ILU (0.919s in the example) reported close the maximum time in the subsequent multiplication by the divergence matrix (which includes nearest neighbor communication), much higher numbers than when you would just use the divergence matrix alone, indicating a synchronization cost while waiting for the slow ILU processors to catch up. Despite large imbalances in local operations, the times for the complete preconditioner show only little variation and do not reveal the imbalances, because there is synchronization immediately before the start (inner products) and at the end (AMG). Even less revealing is the complete solver where the global synchronization points from inner products make times very similar. To see differences, you would need to specifically measure waiting times in MPI, which is beyond our timers. Some parallel performance measuring tools can display exactly such numbers.

To summarize my point: It's certainly revealing interesting things, but you need to focus on  that do not include communication because otherwise the idle times will hide interesting data.

**Comment by tjhei** [comment]
A lightweight tool to instrument MPI calls is IPM (http://ipm-hpc.sourceforge.net/), which I used successfully a couple years ago. It was the only one I tried that did not introduce significant overhead (tau, vampir, ...).  The load balance info (memory, flops, timings, communication) might be interesting, see http://ipm-hpc.sourceforge.net/examples/ex3/ for an example.

I just saw that you can add your own events too: http://ipm-hpc.sourceforge.net/userguide.html 

It would be awesome to see this in action for a test application.

**Comment by bangerth** [comment]
@kronbichler : Yes, I believe that it's hard to see any differences if there are synchronization points somewhere in the timed code. I think your example shows that at least in that code, there are no obvious things that can be addressed.

**Comment by Rombur** [comment]
What do you want to do with it? If you just want to profile the code after a run, you can use hpctoolkit (http://hpctoolkit.org/) I have used it last year. It's really easy to use (you don't need to modify your code), there is almost no overhead, and it can even tell you which function does not scale and how "far" from scalability you are. If you want to get the information during a run to rebalance the mesh on the fly you can get a lot of information with MPI_T_ if you find the documentation...

**Comment by bangerth** [comment]
Maybe when Rene is finally in Texas, I'll ask you to demonstrate its use for all who are interested. I would be interested in running it on a typical 3d case. I have no real idea what we will find but am curious.

**Comment by kronbichler** [comment]
I think #9472 does this.
